### PR TITLE
DM-29032: Switch to C++17.

### PR DESCRIPTION
**Pull request by mwittgen**

### CODE REVIEW

File: python/lsst/sconsUtils/state.py [hunk 1]
```diff
@@ -415,22 +415,22 @@ def ClassifyCc(context):
         env.Append(LINKFLAGS='--coverage')
 
     #
-    # Enable C++14 support (and C99 support for gcc)
+    # Enable C++17 support (and C99 support for gcc)
```
Comment (erykoff) [inline python/lsst/sconsUtils/state.py:418]: Does the "C99" comment mean anything anymore?

**Review by erykoff — APPROVED** [review]

**Comment by TallJimbo** [comment]
Changes look good.  Note that you are welcome to rebase/squash away the reverted commits as long as you are on a branch (and generally should).

As for timing, I think you're okay to make the C++17 switch now, as we seem to be pretty firm about using w_2021_14 for the next release.

But I don't know enough about the macOS deployment targets to comment about the backwards compatibility and timing for that.  If you're confident that you've worked that out with the other stakeholders on e.g. that `#dm-build-problems` slack thread, you can consider this reviewed, but it's not something I'm personally well qualified to either approve or reject.

**Comment by mwittgen** [comment]
I'll squash and leave the MacOS deployment change out for now.

**Comment by timj** [comment]
Updating mac deployment target should be fine. You can link newer against older but you can't link older against newer (and sconsUtils changes will force all the C++ code to be rebuilt anyhow) -- so if it works on Jenkins that's good enough for me.

**Pull request merged**